src/http/mapping_rule/escaping.rs

```diff
@@ -0,0 +1,262 @@
+    const START_RE: &str = r"\A";
+
+    let path_without_dup_fslashes = coalesce_chars(path, '/');
+    let regex_literal = PLACEHOLDER_REGEX
+        .split(path_without_dup_fslashes.as_str())
+        .map(|literal| literal.to_string()) // No regex escaping!
+        .reduce(|mut acc, literal| {
+            acc.push_str(PATH_VALUE_REGEX_S);
+            acc.push_str(literal.as_str());
+            acc
+        })
+        // panic: can't panic because Regex::split always return 1 or more elements in the iterator, so reduce always returns Some()
+        .unwrap();
+
+    let required_capacity = regex_literal
+        .len()
+        .checked_add(START_RE.len())
+        .ok_or_else(|| Error::RegexTooBig)?;
+
+    let mut final_regex = String::with_capacity(required_capacity);
+
+    final_regex.push_str(START_RE);
+    final_regex.push_str(&regex_literal);
+
+    Ok(Regex::new(final_regex.as_str())?)
+}
+
+pub(super) fn split_path_n_qs(s: &str) -> (&str, Option<&str>) {
+    s.find('?')
+        .map(|idx| (&s[..idx], Some(&s[idx + 1..])))
+        .unwrap_or((s, None))
+}
+
+pub(super) fn coalesce_chars(s: &str, coalescing_char: char) -> String {
+    enum Last {
+        Missed,
+        Matched,
+    }
+
+    let mut last = Last::Missed;
+
+    s.chars()
+        .fold(String::with_capacity(s.len()), |mut acc, c| {
+            if c == coalescing_char {
+                if let Last::Missed = last {
+                    acc.push(c);
+                    last = Last::Matched;
+                }
+            } else {
+                acc.push(c);
+                last = Last::Missed;
+            }
+
+            acc
+        })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    mod coalesce_chars {
+        use super::*;
+
+        #[test]
+        fn remove_duped_forward_slashes() -> Result<(), Error> {
+            let pattern = "/a//b///c////d/////e/";
+            assert_eq!(coalesce_chars(pattern, '/').as_str(), "/a/b/c/d/e/");
+
+            Ok(())
+        }
+    }
+
+    mod split_path_n_qs {
+        use super::*;
+
+        #[test]
+        fn strips_question_mark() -> Result<(), Error> {
+            let pattern = "/abc?param1=1&param2=2";
+            let (path, qs) = split_path_n_qs(pattern);
+
+            assert_eq!(path, "/abc");
+            assert_eq!(qs, Some("param1=1&param2=2"));
+
+            Ok(())
+        }
+    }
+
+    mod query_string_regex {
+        use super::*;
+
+        #[test]
+        fn match_qs_parameters() -> Result<(), Error> {
+            let qs_patterns = "fmt={fmt}&hardcoded=1&lang{num}={lang}";
+            let regexes = query_string_regex(qs_patterns)?;
+            let regexes = regexes.iter().map(Regex::as_str).collect::<Vec<_>>();
+
+            assert!(regexes.contains(&format!("fmt={}", QS_VALUE_REGEX_S).as_str()));
+            assert!(regexes.contains(&format!("hardcoded=1").as_str()));
+            assert!(regexes.contains(&format!("lang{r}={r}", r = QS_VALUE_REGEX_S).as_str()));
```

Comment:
This case is not supported by Apicast AFAICS. ([source](https://github.com/3scale/APIcast/blob/e7f6ebbf15d1b053283f0d77ae7176e1c7bfcb44/gateway/src/apicast/mapping_rule.lua#L56-L58))